drivers: audio: Support mimxrt1170_evkb with wm8962 codec in i2s_codec sample

TomasBarakNXP opened the pull request:
- create driver for codec wm8962
- update i2s_codec sample for RT1170 EVKB
- update sai clock in rt11xx.dtsi
- enable SAI mclk in i2s_mcux_sai init, needed for the codec init
- enable playback of sine wave in the i2s codec sample
- format i2s_codec sample code

TomasBarakNXP committed:
drivers: audio: wm8962: Add wm8962 codec support

- create driver for codec wm8962
- add Kconfig, Cmakelist references
- create dts binding

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

butok commented on samples/drivers/i2s/i2s_codec/src/main.c:
```diff
@@ -171,19 +171,14 @@ int main(void)
-					return 0;
-				}
+				/* If not using DMIC, play a sine wave 440Hz */
+				mem_block = (void *)&__16kHz16bit_stereo_sine_pcm;
```
Just curious. Why play anything if the DMIC is not enabled?

TomasBarakNXP replied:
Just to demonstrate the codec TX path - a sine wave can be played to a speaker instead of some random data.

kartben replied:
In any case this behaviour would need to be described in the README, please

dbaluta commented:
Please split this into several patches one patch per logical change.

e.g: - patch for introducing the code for wm8962
       - patch for the main application
       - patch for the dts.

TomasBarakNXP committed:
boards: nxp: mimxrt1170_evk_cm7_B: Add wm8962 codec support

- update mimxrt1170 evkb overlay
- add wm8962 codec on lpi2c5
- add aliases for i2s samples

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

TomasBarakNXP committed:
dts: arm: nxp: Change audio-PLL clock settings on rt11xx

- change SAI clock to 24,576 MHz
  to be compatible with 48kHz sample rate and its derivatives

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

TomasBarakNXP committed:
samples: drivers: i2s: Support sine wave playback in i2s_codec sample

- when not using DMIC, play a 16kHz stereo sine wave using codec

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

dbaluta commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021,2023-2024 NXP Semiconductor INC.
```
Commit subject should say:  `drivers:audio: i2s: Enable SAI mclk`

`Update ... ` is not a real information.

dbaluta commented on dts/arm/nxp/nxp_rt11xx.dtsi:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021,2023-2024 NXP
+ * Copyright 2021,2023-2025 NXP
```
Please avoid formulation like `update..`. Please explain exactly what is done and WHY is done.

decsny commented on dts/arm/nxp/nxp_rt11xx.dtsi:
```diff
@@ -1140,12 +1140,12 @@
+					<&anatop 0 0 32>,
 					<&anatop 0 0 1>,
-					<&anatop 0 0 77>,
-					<&anatop 0 0 100>;
```
im not sure what was the reasoning behind these previous values but maybe, if this is to accomplish something with the board, these values should be changed at the board level

TomasBarakNXP replied:
This change is applicable for the whole rt11xx family. It is setting up the Audio PLL clock to be usable with 48kHz sample rate and its derivatives.

mariopaja commented:
@TomasBarakNXP is there any similarity between wm8962 & wm8960? 
I am asking because I have started implementing wm8960 on rt1064 and would be nice to have some similarities for reference

TomasBarakNXP commented:
> @TomasBarakNXP is there any similarity between wm8962 & wm8960? I am asking because I have started implementing wm8960 on rt1064 and would be nice to have some similarities for reference

Hi @mariopaja there are some similarities, but also some differences. There is an implementation from @hakehuang of the wm8960 [https://github.com/hakehuang/zephyr/commit/38cff1d3115cf80bbc85899131e0d2d6164520c4](here)

mariopaja commented:
@TomasBarakNXP Thank you!

@hakehuang is there any specific reason why your implementation never made it to upstream?

hakehuang commented:
> @hakehuang is there any specific reason why your implementation never made it to upstream?

the work I do before is when the audio is not quite complete. so I keep it in back log.

dbaluta commented on samples/drivers/i2s/i2s_codec/src/main.c:
```diff
@@ -11,21 +11,24 @@
 #if CONFIG_USE_DMIC
-#define NUMBER_OF_CHANNELS  CONFIG_DMIC_CHANNELS
+#define NUMBER_OF_CHANNELS CONFIG_DMIC_CHANNELS
 #else
```
Please do not add style changes in this patch. This makes the real change harder to review. If you really want to do style changes please create a separate patch.

Also, split this patch further.

- one to add sine wave support
- one to add support mimxrt1170 evkb and wm8962 codec

hakehuang commented:
@TomasBarakNXP please update the https://github.com/zephyrproject-rtos/zephyr/blob/19c6240b6865bcb28e1d786d4dcadfb3a02067a0/samples/drivers/i2s/i2s_codec/sample.yaml#L7 add mimxrt1170_evk@B/mimxrt1176/cm7 to the allow list.

TomasBarakNXP committed:
samples: drivers: i2s: Support mimxrt1170 evkb and wm8962 in i2s_codec

- not using DMIC on this board, just play sine wave
- enable sai1 mclk-ouptut for codec init

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

TomasBarakNXP committed:
samples: drivers: i2s: Format i2s_codec sample main.c

- format code according to the coding rules

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

hakehuang reviewed (dismissed):
board testing pass

mmahadevan108 commented:
@anangl , please help review

mmahadevan108 commented:
@TomasBarakNXP please help fix the compliance failures and CI test failures.

TomasBarakNXP commented:
The twister build fails are irrelevant IMHO.
1) CMake build failure: /__w/zephyr/zephyr/tests/bluetooth/classic/sdp_c
- irrelevant to this PR
2) mimxrt1170_evk@B/mimxrt1176/cm7 sample.drivers.i2s.output                          ERROR Build failure
- not supported with this PR

hakehuang commented:
> The twister build fails are irrelevant IMHO.
> 
> 1. CMake build failure: /__w/zephyr/zephyr/tests/bluetooth/classic/sdp_c
> 
> * irrelevant to this PR
> 
> 2. mimxrt1170_evk@B/mimxrt1176/cm7 sample.drivers.i2s.output                          ERROR Build failure
> 
> * not supported with this PR

please rebase to latest zephyr code base, those should be resolved

TomasBarakNXP commented:
Seems that the rebase didn't help.

hakehuang commented:
> Seems that the rebase didn't help.

oh, it is because the upstream CI does not handle the bolbs. We need disable this application in CI just remove the integration for the samples.yaml.

TomasBarakNXP committed:
samples: drivers: i2s: Support mimxrt1170_evkb platform in i2s output

- mimxrt1170_evkb needs CONFIG_DMA_TCD_QUEUE_SIZE=4 which is required
  by the I2S SAI driver
- add test harness for i2s_output

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

hakehuang reviewed (commented):
for i2s_output, pleas help to add, and test OK. Thanks. @TomasBarakNXP 
```   
harness: console
    harness_config:
      type: one_line
      regex:
        - "All I2S blocks written"
```

DerekSnell reviewed (changes requested):
Hi @TomasBarakNXP ,
Sorry to block this now, but I think the MCLK direction should not be hard-coded in the driver.

TomasBarakNXP commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -1125,6 +1125,9 @@ static int i2s_mcux_initialize(const struct device *dev)
 	audio_clock_settings(dev);
 
+	/* enable SAI mclk */
+	enable_mclk_direction(dev, true);
```
Rebased the PR with the requested change in the samples/drivers/i2s/i2s_codec/boards/mimxrt1170_evk_mimxrt1176_cm7_B.overlay

DerekSnell reviewed (dismissed):
Still need to revert the MCLK output change in the i2s_mcux_sai driver.  Thanks

TomasBarakNXP commented:
Build fail in mimxrt1170_evk@A/mimxrt1176/cm7 sample.boards.magic_addr not related to the changes in this PR.

DerekSnell commented:
> Build fail in mimxrt1170_evk@A/mimxrt1176/cm7 sample.boards.magic_addr not related to the changes in this PR.

This PR should resolve the CI build failure once it merges.  https://github.com/zephyrproject-rtos/zephyr/pull/89152

DerekSnell reviewed (dismissed):
Thanks @TomasBarakNXP, 
I have some feedback in case you want to update, but I approve.  Thank you

dleach02 commented:
@anangl this PR is now ready for your approval

TomasBarakNXP committed:
drivers: i2s: mcux_sai: add MCLK direction control in the initialization

Add support for configuring the master clock (MCLK) direction in the
I2S MCUX SAI driver. This allows controlling whether the MCLK pin acts
as an input or output based on device tree configuration, improving
flexibility when interfacing with external audio codecs.

The change leverages the mclk_output property from device tree to
properly set the direction of the MCLK pin during initialization.

Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

hakehuang reviewed (dismissed):
board testing pass on v4.1.0-3789-g2de418b38bc3, but need fix the code compliance issue

hakehuang commented:
board test pass on "v4.1.0-3789-g0cbc89be0d77"

TomasBarakNXP commented:
> ## [![Quality Gate Passed](https://camo.githubusercontent.com/f7c890f34e0efe8252e17e9afd5a85aa22fc9c81b42a0ec3d4821af9f55f250d/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d7061737365642d323070782e706e67)](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=86646) **Quality Gate passed**
> 
> Issues ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [1 New issue](https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=86646&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) ![](https://camo.githubusercontent.com/86e5c40907bf9e8933fe6bcf185f7076622a76a3bef4b6d2b5fd583959409aef/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f61636365707465642d313670782e706e67) [0 Accepted issues](https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=86646&issueStatuses=ACCEPTED)
> 
> Measures ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=zephyrproject-rtos_zephyr&pullRequest=86646&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0.0% Coverage on New Code](https://sonarcloud.io/component_measures?id=zephyrproject-rtos_zephyr&pullRequest=86646&metric=new_coverage&view=list) ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [13.2% Duplication on New Code](https://sonarcloud.io/component_measures?id=zephyrproject-rtos_zephyr&pullRequest=86646&metric=new_duplicated_lines_density&view=list)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=86646)

Addressed the findings.

hakehuang commented:
board testing pass on v4.1.0-3789-g80ed010a34d6
